### PR TITLE
Cleaned up FCMRegistrationService 

### PR DESCRIPTION
**Pull request by anitaa1990**
Step 1 of #4525. This PR cleans up the FCM token registration process so that it can be unit testable.

### Changes
- Created a new class `NotificationRegistrationHandler` that handles the logic after getting the new FCM token. 
- Removed the deprecated `InstanceIDService` and instead use the recommended `onNewToken()` method in `FCMMessagingService`.
- Cleaned up `FCMRegistrationService` to utilise `NotificationRegistrationHandler`.
- Added unit tests for `NotificationRegistrationHandler`.

### Testing
There are no user facing changes. Testing the app to verify that you are able to receive new order, new review and Zendesk notifications when the app is in the foreground, background or closed completely, should be sufficient. 

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMMessageService.kt [hunk 1]
```diff
@@ -18,6 +18,17 @@ class FCMMessageService : FirebaseMessagingService() {
...
+
+    /**
+     * Called if the FCM registration token is updated. This may occur if the security of
+     * the previous token had been compromised. Note that this is called when the
+     * FCM registration token is initially generated so this is where you would retrieve the token.
+     */
+    override fun onNewToken(newToken: String) {
+        WooLog.d(T.NOTIFS, "Sending FCM token to our remote services: $newToken")
+        notificationRegistrationHandler.onNewFCMTokenReceived(newToken)
+    }
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMMessageService.kt:31]: 👏 Great! That's a nice change with moving `onNewToken` here

File: WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMRegistrationIntentService.kt [hunk 2]
```diff
@@ -3,73 +3,42 @@ package com.woocommerce.android.push
...
-import org.wordpress.android.fluxc.Dispatcher
-import org.wordpress.android.fluxc.generated.NotificationActionBuilder
-import org.wordpress.android.fluxc.store.AccountStore
-import org.wordpress.android.fluxc.store.NotificationStore
-import org.wordpress.android.fluxc.store.NotificationStore.NotificationAppKey.WOOCOMMERCE
-import org.wordpress.android.fluxc.store.NotificationStore.RegisterDevicePayload
 import javax.inject.Inject
 
 @AndroidEntryPoint
 class FCMRegistrationIntentService : JobIntentService() {
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMRegistrationIntentService.kt:13]: ❓ As we start to use `FirebaseMessagingService#onNewToken`, is `FCMRegistrationIntentService` needed at all? WDYT about removing it?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/push/NotificationRegistrationHandler.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import com.woocommerce.android.util.PreferencesWrapper
+import org.wordpress.android.fluxc.Dispatcher
+import org.wordpress.android.fluxc.generated.NotificationActionBuilder
+import org.wordpress.android.fluxc.store.AccountStore
+import org.wordpress.android.fluxc.store.NotificationStore
+import org.wordpress.android.fluxc.store.NotificationStore.RegisterDevicePayload
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/push/NotificationRegistrationHandler.kt]: 🔧 Maybe we could remove `@Singleton` from here as it doesn't keep any state?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/push/NotificationRegistrationHandler.kt [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    private val notificationStore: NotificationStore, // Required to ensure the notificationStore is initialized
+    private val preferencesWrapper: PreferencesWrapper,
+    private val selectedSite: SelectedSite
+) {
+    fun onNewFCMTokenReceived(token: String) {
+        // Register for WordPress.com notifications only if user is logged in
+        if (accountStore.hasAccessToken()) {
+            preferencesWrapper.setFCMToken(token)
+
+            // The site is set to null to ensure we get notifications for all stores the user is logged in
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/push/NotificationRegistrationHandler.kt]: ❓ I'm not sure if I got that comment, should `site` be `null` in line 30?

File: WooCommerce/src/test/kotlin/com/woocommerce/android/push/NotificationRegistrationHandlerTest.kt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Test
+    fun `new fcm token is registered successfully only if user is logged in`() {
+        doReturn(true).whenever(accountStore).hasAccessToken()
+
+        val fcmToken = "123456"
+        notificationRegistrationHandler.onNewFCMTokenReceived(token = fcmToken)
+
+        verify(preferencesWrapper, atLeastOnce()).setFCMToken(any())
+        verify(dispatcher, atLeastOnce()).dispatch(actionCaptor.capture())
```
Comment (wzieba) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/push/NotificationRegistrationHandlerTest.kt]: 🔧 I think those `argumentCaptor` are not needed, WDYT?

```suggestion
        verify(dispatcher, never()).dispatch(any())
    }

    @Test
    fun `new fcm token is registered successfully only if user is logged in`() {
        doReturn(true).whenever(accountStore).hasAccessToken()

        val fcmToken = "123456"
        notificationRegistrationHandler.onNewFCMTokenReceived(token = fcmToken)

        verify(preferencesWrapper, atLeastOnce()).setFCMToken(any())
        verify(dispatcher, atLeastOnce()).dispatch(any())
```

**Review by wzieba — COMMENTED** [review]
Thank you @anitaa1990 for working on this! I've left some comments, WDYT?

I've tested this change with clean install of release version too and everything seems to work correct 👍

**Review by wzieba — APPROVED** [review]
I'm leaving approve as none of those comments should be blocking merge 🙌

File: WooCommerce/src/main/kotlin/com/woocommerce/android/push/NotificationRegistrationHandler.kt (hunk 4 above)
Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/push/NotificationRegistrationHandler.kt]: So I had first set the `site` to null. This is needed to support notifications for multiple sites, which is not the scope of this PR. So changed it to support a single site but forgot to remove the comment 🤦‍♀️ . Fixed in c12b148

File: WooCommerce/src/main/kotlin/com/woocommerce/android/push/NotificationRegistrationHandler.kt (hunk 3 above)
Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/push/NotificationRegistrationHandler.kt]: Fixed in c12b148 👍

File: WooCommerce/src/main/kotlin/com/woocommerce/android/push/NotificationRegistrationHandler.kt (hunk 4 above)
Comment (wzieba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/push/NotificationRegistrationHandler.kt]: ok gotcha, thanks!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMRegistrationIntentService.kt (hunk 2 above)
Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMRegistrationIntentService.kt:13]: So what is happening is that we need to re-register the fcm token after a user logs into the app or switches stores. And we seem to be using `FCMRegistrationIntentService` for that. This fetches the latest fcm token and sends an api request to register this token to the current site. Once we support push notifications for multiple stores, we can definitely remove it but I'll leave it in for now 👍 WDYT?

Comment (wzieba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMRegistrationIntentService.kt:13]: Right, gotcha - I didn't think of case with switching stores. Thank you for explanation!

> Once we support push notifications for multiple stores, we can definitely remove it but I'll leave it in for now 👍 WDYT?

Sure, sounds good!

**Comment by anitaa1990** [comment]
Fixed the changes in the feedback @wzieba! Thank you for the review! Ready for another round 👍

**Review by wzieba — APPROVED** [review]
Thank you @anitaa1990! LGTM :shipit:

**Pull request merged**